client/services/LocalStorageService.js

```diff
@@ -0,0 +1,196 @@
+     */
+    var getLocalStorageKeyForCode = function(questionId, language) {
+      return questionId + ":" + language + ":code";
+    };
+
+    /**
+     * Returns the local storage key for the feedback.
+     *
+     * @param {string} questionId
+     * @param {string} language
+     * @returns {string}
+     */
+    var getLocalStorageKeyForFeedback = function(questionId, language) {
+      return questionId + ":" + language + ":feedback";
+    };
+
+    return {
+      /**
+       * Checks if the local storage is available.
+       *
+       * @returns {boolean}
+       */
+      isAvailable: function() {
+        return localStorageIsAvailable;
+      },
+
+      /**
+       * Stores the code into local storage - so long as local storage is
+       * available.
+       *
+       * @param {string} questionId
+       * @param {string} code
+       * @param {string} language
+       */
+      storeCode: function(questionId, code, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        // Add version to the code, currently only 1
+        // TODO(talee): Add in check for the correct version number
+        var codeWithVersion = "1:" + code;
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        localStorage.setItem(localStorageKey, codeWithVersion);
+      },
+
+      /**
+       * Loads the code from local storage if local storage is available.
+       * If local storage is not available, then it returns null.
+       *
+       * @param {string} questionId
+       * @param {string} language
+       * @returns {string|null}
+       */
+      loadStoredCode: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return null;
+        }
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        var storedCode = localStorage.getItem(localStorageKey);
+
+        if (storedCode === null) {
+          return null;
+        }
+        // TODO(talee): Check versioning when we start having different versions
+        storedCode = storedCode.substring(storedCode.indexOf(':') + 1);
+
+        return storedCode;
+      },
+
+      /**
+       * Clears the local storage such that there is no longer any code
+       * stored there.
+       *
+       * @param {string} questionId
+       * @param {string} language
+       */
+      clearLocalStorageCode: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        localStorage.removeItem(localStorageKey);
+      },
+
+      storeFeedback: function(questionId, feedback, reinforcement, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        var localStorageKey = getLocalStorageKeyForFeedback(
+          questionId, language);
+        var feedbackWithReinforcement = {};
```

Comment:
I think this is the point where you call the toDict() methods of the relevant classes.